src/ripple/protocol/AcctRoot.h

```diff
@@ -0,0 +1,354 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2021 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_PROTOCOL_ACCT_ROOT_H_INCLUDED
+#define RIPPLE_PROTOCOL_ACCT_ROOT_H_INCLUDED
+
+#include <ripple/protocol/LedgerEntryWrapper.h>
+#include <ripple/protocol/STAccount.h>
+#include <ripple/protocol/STAmount.h>
+
+namespace ripple {
+
+template <bool Writable>
+class AcctRootImpl final : public LedgerEntryWrapper<Writable>
+{
+private:
+    using Base = LedgerEntryWrapper<Writable>;
+    using SleT = typename Base::SleT;
+    using Base::wrapped_;
+
+    // This constructor is private so only the factory functions can
+    // construct an AcctRootImpl.
+    AcctRootImpl(std::shared_ptr<SleT>&& w) : Base(std::move(w))
+    {
+    }
+
+    // Friend declarations of factory functions.
+    //
+    // For classes that contain factories we must declare the entire class
+    // as a friend unless the class declaration is visible at this point.
+    friend class ReadView;
+    friend class ApplyView;
+
+public:
+    AcctRootImpl(AcctRootImpl const&) = default;
+    AcctRootImpl(AcctRootImpl&&) = default;
+
+    AcctRootImpl&
+    operator=(AcctRootImpl const& rhs)
+    {
+        Base::operator=(rhs);
+        return *this;
+    }
+
+    AcctRootImpl&
+    operator=(AcctRootImpl&& rhs)
+    {
+        Base::operator=(std::move(rhs));
+        return *this;
+    }
+
+    ~AcctRootImpl() = default;
+
+    // Conversion operator from AcctRootImpl<true> to AcctRootImpl<false>.
+    operator AcctRootImpl<true>() const
+    {
+        return AcctRootImpl<false>(
+            std::const_pointer_cast<std::shared_ptr<STLedgerEntry const>>(
```

Comment:
why isn't the type of the shared_ptr `AcctRootImpl` ? Wouldn't that provide a stronger type guarentee?

Reply:
okay, my bad. got it 👍